File: api-clients/elixir/lib/otreeba_open_cannabis_api/api/flowers.ex
defmodule OtreebaOpenCannabisAPI.Api.Flowers do
  @moduledoc """
  Documentation for OtreebaOpenCannabisAPI.Api.Flowers.
  """

  use Tesla

  plug Tesla.Middleware.BaseUrl, "https://api.otreeba.com/v1"
  plug Tesla.Middleware.JSON

  @doc """
  Find flower by Open Cannabis Product Code (OCPC).

  Returns a single flower.
  """
  def get_flower_by_ocpc(ocpc) do
    method = [method: :get]
    url = [url: "/flowers/#{ocpc}"]
    query_params = []
    header_params = []
    body_params = []
    form_params = []
    params = query_params ++ header_params ++ body_params ++ form_params
    opts = []
    options = method ++ url ++ params ++ opts

    request(options)
  end

  @doc """
  Get a list of all current flowers.

  Returns a paginated list of flowers.
  """
  def get_flowers(page, count, sort) do
    method = [method: :get]
    url = [url: "/flowers"]
    query_params = [query: [{:"page", page}, {:"count", count}, {:"sort", sort}]]
    header_params = []
    body_params = []
    form_params = []
    params = query_params ++ header_params ++ body_params ++ form_params
    opts = []
    options = method ++ url ++ params ++ opts

    request(options)
  end
end

File: api-clients/clojure/src/otreeba_open_cannabis_api/api/flowers.clj
(ns otreeba-open-cannabis-api.api.flowers
  (:require [otreeba-open-cannabis-api.core :refer [call-api check-required-params with-collection-format]])
  (:import (java.io File)))

(defn get-flower-by-ocpc-with-http-info
  "Find flower by Open Cannabis Product Code (OCPC).
  Returns a single flower."
  [ocpc ]
  (call-api "/flowers/{ocpc}" :get
            {:path-params   {"ocpc" ocpc }
             :header-params {}
             :query-params  {}
             :form-params   {}
             :content-types ["application/json"]
             :accepts       ["application/json"]
             :auth-names    []}))

(defn get-flower-by-ocpc
  "Find flower by Open Cannabis Product Code (OCPC).
  Returns a single flower."
  [ocpc ]
  (:data (get-flower-by-ocpc-with-http-info ocpc)))

(defn get-flowers-with-http-info
  "Get a list of all current flowers.
  Returns a paginated list of flowers."
  ([] (get-flowers-with-http-info nil))
  ([{:keys [page count sort ]}]
   (call-api "/flowers" :get
             {:path-params   {}
              :header-params {}
              :query-params  {"page" page "count" count "sort" sort }
              :form-params   {}
              :content-types ["application/json"]
              :accepts       ["application/json"]
              :auth-names    []})))

(defn get-flowers
  "Get a list of all current flowers.
  Returns a paginated list of flowers."
  ([] (get-flowers nil))
  ([optional-params]
   (:data (get-flowers-with-http-info optional-params))))

